### PR TITLE
Generate DWARF address ranges for faster lookups

### LINKED ISSUE #45246: Consider emitting DWARF aranges
I recently discovered that the elfutils `eu-addr2line` [doesn't resolve locations for Rust functions](https://sourceware.org/bugzilla/show_bug.cgi?id=22288).  It turns out that they are expecting `.debug_aranges` to be valid and complete, and even treat its absence as if there are no addresses covered at all.  But `rustc` doesn't emit aranges with its debuginfo, or at least not by default.

If you see a `.debug_aranges` section in your rust binaries today, that's probably only from objects compiled by GCC, like jemalloc.  Try `eu-readelf -waranges` to see what CUs and symbols are covered.  Thus even when `.debug_aranges` does exist, it's not necessarily complete.

Clang doesn't emit aranges by default either, but it does have the `-gdwarf-aranges` option to do so.  This turns into the LLVM option `-generate-arange-section`.  So it turns out that Rust can have aranges too, using `rustc -Cllvm-args=-generate-arange-section`.  Great!

It would be nice if this were the default though.  The data is usually not very big, but it can let tools map an address to the right CU very quickly, reducing the slower DIE traversals.

### PR DESCRIPTION
**Pull request by cuviper**
This adds a new option `-Zgenerate-arange-section`, enabled by default,
corresponding to LLVM's `-generate-arange-section`. This creates a
`.debug_aranges` section with DWARF address ranges, which some tools
depend on to optimize address lookups (elfutils [22288], [25173]).

This only has effect when debuginfo is enabled, and the additional data
is small compared to the other debug sections. For example, libstd.so
with full debuginfo is about 11MB, with just 61kB in aranges.

[22288]: https://sourceware.org/bugzilla/show_bug.cgi?id=22288
[25173]: https://sourceware.org/bugzilla/show_bug.cgi?id=25173

Closes #45246.
r? @michaelwoerister

### CODE REVIEW

**Comment by cuviper** [comment]
We could of course choose a different option name, and I wasn't sure if it should be `TRACKED`.

**Comment by Mark-Simulacrum** [comment]
Seems like it should be tracked, since we presumably want to rerun codegen (and debuginfo-gen) if it changes.

How is the performance side of this? We can check somewhat on perf I guess...

@bors try @rust-timer queue

File: src/librustc/session/config.rs [hunk 1]
```diff
@@ -1331,6 +1331,8 @@ options! {DebuggingOptions, DebuggingSetter, basic_debugging_options,
         "for every macro invocation, print its name and arguments"),
     debug_macros: bool = (false, parse_bool, [TRACKED],
         "emit line numbers debug info inside macros"),
+    generate_arange_section: bool = (true, parse_bool, [UNTRACKED],
```
Comment (michaelwoerister) [inline src/librustc/session/config.rs]: This should be `[TRACKED]` as the setting affects output files and we'd want to rebuild them if the setting changes.

**Review by michaelwoerister — COMMENTED** [review]
Thanks, @cuviper! Looks good to me. Let's get this merged asap so we get some testing while still early in the current cycle.

File: src/librustc/session/config.rs (hunk 1 above)
Comment (cuviper) [reply src/librustc/session/config.rs]: OK, I changed it.

**Comment by michaelwoerister** [comment]
Let's keep this change bisectable:
@bors rollup never

**Comment by cuviper** [comment]
So it was rolled up anyway -- I think the bot syntax is "rollup=never".

**Pull request merged**